lightning/src/ln/channel.rs

```diff
@@ -1710,9 +1778,54 @@ impl<ChanSigner: ChannelKeys> Channel<ChanSigner> {
 				removed_outbound_total_msat += htlc.amount_msat;
 			}
 		}
-		if htlc_inbound_value_msat + msg.amount_msat + self.value_to_self_msat > (self.channel_value_satoshis - Channel::<ChanSigner>::get_remote_channel_reserve_satoshis(self.channel_value_satoshis)) * 1000 + removed_outbound_total_msat {
-			return Err(ChannelError::Close("Remote HTLC add would put them under their reserve value"));
+
+		if self.channel_outbound {
+			// Check that they won't violate our local required channel reserve by adding this HTLC.
+
+			// +1 for this HTLC.
+			let local_fee_cost_msat = self.next_local_commit_tx_fee_msat(1);
+			if self.value_to_self_msat < self.local_channel_reserve_satoshis * 1000 + local_fee_cost_msat {
+				return Err(ChannelError::Ignore("Cannot receive value that would put us under local channel reserve value"));
+			}
+		}
+
+		let pending_value_to_self_msat =
+			self.value_to_self_msat + htlc_inbound_value_msat - removed_outbound_total_msat;
+		let pending_remote_value_msat =
+			self.channel_value_satoshis * 1000 - pending_value_to_self_msat;
+		if pending_remote_value_msat < msg.amount_msat {
+			return Err(ChannelError::Close("Remote HTLC add would overdraw remaining funds"));
+		}
+
+		let remote_fee_cost_msat = if self.channel_outbound { 0 } else {
+			// +1 for this HTLC.
+			self.next_remote_commit_tx_fee_msat(1)
+		};
+		if pending_remote_value_msat - msg.amount_msat < remote_fee_cost_msat {
+			return Err(ChannelError::Close("Remote HTLC add would not leave enough to pay for fees"));
+		};
+
+		let channel_reserve_msat =
+			Channel::<ChanSigner>::get_remote_channel_reserve_satoshis(self.channel_value_satoshis) * 1000;
+		if pending_remote_value_msat - msg.amount_msat - remote_fee_cost_msat < channel_reserve_msat {
+			return Err(ChannelError::Close("Remote HTLC add would put them under remote reserve value"));
 		}
+
+		if !self.channel_outbound {
+			// `+1` for this HTLC, `2 *` and `+1` fee spike buffer we keep for the remote (this deviates from the spec
+			// but should help protect us from stuck channels).
+			let remote_fee_cost_incl_stuck_buffer_msat = 2 * self.next_remote_commit_tx_fee_msat(1 + 1);
+			if pending_remote_value_msat - msg.amount_msat - channel_reserve_msat < remote_fee_cost_incl_stuck_buffer_msat {
+				if let Some(reason) = temp_chan_fail_reason {
+					pending_forward_state = PendingHTLCStatus::Fail(HTLCFailureMsg::Relay(msgs::UpdateFailHTLC {
+						channel_id: msg.channel_id,
+						htlc_id: msg.htlc_id,
+						reason,
+					}));
+				}
```

Comment:
Why doesn't this cause an `Err` to be returned? If it could be then, perhaps `temp_chan_fail_reason` could be left out of this method entirely.

Reply:
Because atm there's no way to return an error without the channel closing (even ChanError::Ignore will cause the channel to close), see https://github.com/rust-bitcoin/rust-lightning/pull/577#discussion_r417595712

This was the least invasive way I found to accomplish the goal prescribed in the comment^ but I agree it's not super ideal.

Reply:
Hm, maybe I could add a new `ChannelError::FailHTLC` that somehow indicates what pending_forward_state we want the HTLC to fail with, and return that from `update_add_htlc`

Reply:
The reason I implemented https://github.com/rust-bitcoin/rust-lightning/pull/577#discussion_r417595712 this way is because the `Channel` has no way of constructing a channel update for use in the `temporary_channel_failure` onion error. Empowering the Channel to construct a channel update would require giving it a pointer to a `FeeEstimator`, and the node's secret key (this option doesn't seem too much easier to follow than the current one).

Alternatively, I could make the leave the `data` field empty in `HTLCFailReason::Reason { failure_code, data }`, wait til the failure bubbles up in the ChannelManager, manually check to see if `failure code == temp chan failure && data field is blank == true` and if so, add the channel update then. This seems more error prone and equally hacky?

I could be missing other options.. Lmk what you think.

Reply:
Another option could be to pass a closure for constructing the reason. It would at very least allow you to only create it when necessary.

Still might be worth refactoring in some other way, but I'll have to stare at it a bit more first.

Reply:
Sg, let me know what you conclude

Reply:
Here is the closure-based approach:

```diff
diff --git a/lightning/src/ln/channel.rs b/lightning/src/ln/channel.rs
index 511f70e..a6b5f89 100644
--- a/lightning/src/ln/channel.rs
+++ b/lightning/src/ln/channel.rs
@@ -1733,7 +1733,8 @@ impl<ChanSigner: ChannelKeys> Channel<ChanSigner> {
                self.commit_tx_fee_msat(their_acked_htlcs + addl_htlcs)
        }
 
-       pub fn update_add_htlc(&mut self, msg: &msgs::UpdateAddHTLC, mut pending_forward_state: PendingHTLCStatus, temp_chan_fail_reason: Option<msgs::OnionErrorPacket>) -> Result<(), ChannelError> {
+       pub fn update_add_htlc<F>(&mut self, msg: &msgs::UpdateAddHTLC, mut pending_forward_state: PendingHTLCStatus, create_onion_error_packet: F) -> Result<(), ChannelError>
+       where F: for<'a, 'b> Fn(&'a Self, &'b PendingHTLCStatus) -> Option<msgs::OnionErrorPacket> {
                if (self.channel_state & (ChannelState::ChannelFunded as u32 | ChannelState::RemoteShutdownSent as u32)) != (ChannelState::ChannelFunded as u32) {
                        return Err(ChannelError::Close("Got add HTLC message when channel was not in an operational state"));
                }
@@ -1816,7 +1817,8 @@ impl<ChanSigner: ChannelKeys> Channel<ChanSigner> {
                        // but should help protect us from stuck channels).
                        let remote_fee_cost_incl_stuck_buffer_msat = 2 * self.next_remote_commit_tx_fee_msat(1 + 1);
                        if pending_remote_value_msat - msg.amount_msat - chan_reserve_msat < remote_fee_cost_incl_stuck_buffer_msat {
-                               if let Some(reason) = temp_chan_fail_reason {
+                               let onion_error_packet = create_onion_error_packet(self, &pending_forward_state);
+                               if let Some(reason) = onion_error_packet {
                                        pending_forward_state = PendingHTLCStatus::Fail(HTLCFailureMsg::Relay(msgs::UpdateFailHTLC {
                                                channel_id: msg.channel_id,
                                                htlc_id: msg.htlc_id,
diff --git a/lightning/src/ln/channelmanager.rs b/lightning/src/ln/channelmanager.rs
index 0ed3187..ec642a3 100644
--- a/lightning/src/ln/channelmanager.rs
+++ b/lightning/src/ln/channelmanager.rs
@@ -2514,19 +2514,21 @@ impl<ChanSigner: ChannelKeys, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref>
                                                }));
                                        }
                                }
-                               let temp_chan_failure_reason = match pending_forward_info {
-                                       PendingHTLCStatus::Forward(PendingHTLCInfo { incoming_shared_secret, .. }) => {
-                                               let chan_update = self.get_channel_update(chan.get()).unwrap();
-                                               Some(onion_utils::build_first_hop_failure_packet(&incoming_shared_secret, 0x1000|7, &{
-                                                       let mut res = Vec::with_capacity(8 + 128);
-                                                       res.extend_from_slice(&byte_utils::be16_to_array(chan_update.contents.flags));
-                                                       res.extend_from_slice(&chan_update.encode_with_len()[..]);
-                                                       res
-                                               }[..]))
-                                       },
-                                       _ => None,
+                               let create_onion_error_packet = |chan: &Channel<ChanSigner>, pending_forward_info: &PendingHTLCStatus| {
+                                       match pending_forward_info {
+                                               &PendingHTLCStatus::Forward(PendingHTLCInfo { ref incoming_shared_secret, .. }) => {
+                                                       let chan_update = self.get_channel_update(chan).unwrap();
+                                                       Some(onion_utils::build_first_hop_failure_packet(incoming_shared_secret, 0x1000|7, &{
+                                                               let mut res = Vec::with_capacity(8 + 128);
+                                                               res.extend_from_slice(&byte_utils::be16_to_array(chan_update.contents.flags));
+                                                               res.extend_from_slice(&chan_update.encode_with_len()[..]);
+                                                               res
+                                                       }[..]))
+                                               },
+                                               _ => None,
+                                       }
                                };
-                               try_chan_entry!(self, chan.get_mut().update_add_htlc(&msg, pending_forward_info, temp_chan_failure_reason), channel_state, chan);
+                               try_chan_entry!(self, chan.get_mut().update_add_htlc(&msg, pending_forward_info, create_onion_error_packet), channel_state, chan);
                        },
                        hash_map::Entry::Vacant(_) => return Err(MsgHandleErrInternal::send_err_msg_no_close("Failed to find corresponding channel", msg.channel_id))
                }

```

The benefit is that error packet is only constructed as needed. It also keeps all the error construction together as you already have done. You may be able to iterate on it to create the entire `PendingHTLCStatus::Fail` rather than just the reason. May make the code in channel.rs a bit simpler that way. Not sure.

Regarding some other refactoring, that may be a bit more involved. I don't have a great sense as to how go about this. I find it a little strange that in the other failure cases (i.e. when `ChannelManager` sees the channel is not usable), we still call `update_add_htlc` which may then fail for an entirely different reason and not use the passed `pending_forward_info`. This isn't very intuitive to me, so it seems like there is room for improvement down the road.

Reply:
Awesome thanks, applied the diff :) (Btw, `git apply jeffdiff.diff` didn't work, but `git apply --ignore-space-change --ignore-whitespace` did work.. not sure why).

Yeah, agreed it's a bit funky. Though, if we come across a reason to fail an _HTLC_, then subsequently come across a reason to fail the entire _channel_, it would make sense to ignore the original reason and just close the channel I guess

Reply:
Ah, so I was trying to diff your version and mine and couldn't figure out why it reported diffs even though everything _looked_ the same. Turns out, I copied my version from the output of `git diff` but the tabs were somehow converted to spaces in the process. :P Think you'll have to manually convert those. Sorry about that.

Reply:
Ah, that would do it. 

Btw, w/r/t reviewing the tests -- some of those tests were added because the `do_channel_reserve_test` function was difficult to update. So, I split off some parts of what it was testing into their own individual tests, to maintain the same level of coverage but ease updates.

Reply:
Looks like they are still spaces, just less of them. :)